client/lib/wpcom-undocumented/lib/undocumented.js

```diff
@@ -787,9 +791,30 @@ Undocumented.prototype.cart = function( cartKey, method, data, fn ) {
 		method = 'GET';
 		data = {};
 	}
+
+	if ( method === 'POST' ) {
+		// If no coupon and we aren't explicitly removing the coupon,
+		// then use the coupon from query string if one is available.
+		data = data || {};
+
+		if (
+			! data.coupon &&
+			! data.is_coupon_applied &&
+			! data.is_coupon_removed &&
+			typeof document !== 'undefined'
+		) {
+			const coupon = getQueryCoupon();
```

Comment:
note that right here we're strongly coupling the URL in the browser with the API call we're making to the backend. this isn't obvious at all and is going to be an additional layer of opacity when trying to test this module.

one good way we can get around this is by trapping the URL query argument somewhere else such as in the navigation/routing middleware and storing it in state then passing that state along into this function. if this function took a coupons parameter then it would all be explicit and clear.

Reply:
Thanks! I updated this and left notes in https://github.com/Automattic/wp-calypso/pull/25433#discussion_r194726451